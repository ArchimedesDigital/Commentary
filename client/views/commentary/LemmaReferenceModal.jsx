import RaisedButton from "material-ui/RaisedButton";

LemmaReferenceModal = React.createClass({

	propTypes: {
		visible: React.PropTypes.bool,
		top: React.PropTypes.number,
		left: React.PropTypes.number,
		work: React.PropTypes.string,
		subwork: React.PropTypes.number,
		lineFrom: React.PropTypes.number,
		lineTo: React.PropTypes.number,
		closeLemmaReference: React.PropTypes.func,
	},

	getInitialState() {
		return {
			selectedLemmaEdition: {lines: []},

		};
	},

	mixins: [ReactMeteorData],

	getMeteorData() {
		const lemmaQuery = {
			'work.slug': this.props.work,
			'subwork.n': this.props.subwork,
			'text.n': {
				$gte: this.props.lineFrom,
			},
		};
		let lemmaText = [];

		if (this.props.lineTo) {
			lemmaQuery['text.n'].$lte = this.props.lineTo;
		} else {
			lemmaQuery['text.n'].$lte = this.props.lineFrom;
		}

		const textHandle = Meteor.subscribe('textNodes', lemmaQuery);
		if (textHandle.ready()) {
			// console.log('lemmaQuery', lemmaQuery);
			const textNodes = TextNodes.find(lemmaQuery).fetch();
			const editions = [];

			let textIsInEdition = false;
			textNodes.forEach(function (textNode) {
				textNode.text.forEach(function (text) {
					textIsInEdition = false;

					editions.forEach(function (edition) {
						if (text.edition.slug === edition.slug) {
							edition.lines.push({
								html: text.html,
								n: text.n,
							});
							textIsInEdition = true;
						}
					});

					if (!textIsInEdition) {
						editions.push({
							title: text.edition.title,
							slug: text.edition.slug,
							lines: [
								{
									html: text.html,
									n: text.n,
								},
							],
						});
					}
				});
			});

			lemmaText = editions;
		}

		return {
			lemmaText,
		};
	},

	componentDidUpdate() {
		// console.log('lemmaReferenceModal', this.props);
		if (this.data.lemmaText.length && this.state.selectedLemmaEdition.lines.length === 0) {
			this.setState({
				selectedLemmaEdition: this.data.lemmaText[0],
			});
		} else if (!this.props.visible && this.state.selectedLemmaEdition.lines.length) {
			this.setState({
				selectedLemmaEdition: {lines: []},
			});
		}
	},

	toggleEdition(editionSlug) {
		if (this.state.selectedLemmaEdition.slug !== editionSlug) {
			let newSelectedEdition = {};
			this.data.lemmaText.forEach(function (edition) {
				if (edition.slug === editionSlug) {
					newSelectedEdition = edition;
				}
			});

			this.setState({
				selectedLemmaEdition: newSelectedEdition,
			});
		}
	},

	render() {
		const self = this;
		const lemmaText = this.data.lemmaText;
		const styles = {
			lemmaReferenceModal: {
				top: this.props.top,
				left: this.props.left,

			},

		};
		const hasLemma = ~this.state.selectedLemmaEdition.lines.length;

		return (
			<div
				className={'lemma-reference-modal' + ((this.props.visible && hasLemma) ? ' lemma-reference-modal-visible' : '')}
				style={styles.lemmaReferenceModal}
			>
				<article className="comment	lemma-comment paper-shadow ">

					<div className="lemma-reference-text">
						{this.state.selectedLemmaEdition.lines.map(function (line, i) {
							return (<p
								key={i}
								className="lemma-text"
								dangerouslySetInnerHTML={{__html: line.html}}
							/>);
						})}
					</div>

					<div className="edition-tabs tabs">
						{lemmaText.map(function (lemmaTextEdition, i) {
							const lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

							return (<RaisedButton
								key={i}
								label={lemmaEditionTitle}
								data-edition={lemmaTextEdition.title}
								className={self.state.selectedLemmaEdition.slug === lemmaTextEdition.slug ? 'edition-tab tab selected-edition-tab' : 'edition-tab tab'}
								onClick={self.toggleEdition.bind(null, lemmaTextEdition.slug)}
							/>);
						})}
					</div>

					<i
						className="mdi mdi-close paper-shadow"
						onClick={this.props.closeLemmaReference}
					/>
				</article>

			</div>

		);
	},

});
