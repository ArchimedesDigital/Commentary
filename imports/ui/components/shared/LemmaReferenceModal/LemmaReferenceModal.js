import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import RaisedButton from 'material-ui/RaisedButton';
import TextNodes from '/imports/api/collections/textNodes';

LemmaReferenceModal = React.createClass({

	propTypes: {
		visible: React.PropTypes.bool,
		top: React.PropTypes.number.isRequired,
		left: React.PropTypes.number.isRequired,
		work: React.PropTypes.string.isRequired,
		subwork: React.PropTypes.number.isRequired,
		lineFrom: React.PropTypes.number.isRequired,
		lineTo: React.PropTypes.number,
		closeLemmaReference: React.PropTypes.func.isRequired,
	},

	childContextTypes: {
		muiTheme: React.PropTypes.object.isRequired,
	},

	mixins: [ReactMeteorData],

	getInitialState() {
		return {
			selectedLemmaEditionIndex: 0,

		};
	},

	getChildContext() {
		return { muiTheme: getMuiTheme(baseTheme) };
	},

	getMeteorData() {
		const lemmaQuery = {
			'work.slug': this.props.work,
			'subwork.n': this.props.subwork,
			'text.n': {
				$gte: this.props.lineFrom,
			},
		};
		let lemmaText = [];

		if (this.props.lineTo) {
			lemmaQuery['text.n'].$lte = this.props.lineTo;
		} else {
			lemmaQuery['text.n'].$lte = this.props.lineFrom;
		}

		const textHandle = Meteor.subscribe('textNodes', lemmaQuery);
		if (textHandle.ready()) {
			const textNodes = TextNodes.find(lemmaQuery).fetch();
			const editions = [];

			let textIsInEdition = false;
			textNodes.forEach((textNode) => {
				textNode.text.forEach((text) => {
					textIsInEdition = false;

					editions.forEach((edition) => {
						if (text.edition.slug === edition.slug) {
							edition.lines.push({
								html: text.html,
								n: text.n,
							});
							textIsInEdition = true;
						}
					});

					if (!textIsInEdition) {
						editions.push({
							title: text.edition.title,
							slug: text.edition.slug,
							lines: [
								{
									html: text.html,
									n: text.n,
								},
							],
						});
					}
				});
			});

			lemmaText = editions;
		}

		return {
			lemmaText,
		};
	},

	toggleEdition(editionSlug) {
		if (this.data.lemmaText.length) {
			if (this.data.lemmaText[this.state.selectedLemmaEditionIndex].slug !== editionSlug) {
				let newSelectedEditionIndex = 0;
				this.data.lemmaText.forEach((edition, index) => {
					if (edition.slug === editionSlug) {
						newSelectedEditionIndex = index;
					}
				});

				this.setState({
					selectedLemmaEditionIndex: newSelectedEditionIndex,
				});
			}
		}
	},

	toggleHighlighting() {
		this.setState({
			highlightingVisible: !this.state.highlightingVisible,
		});
	},

	render() {
		const self = this;
		const lemmaText =
			this.data.lemmaText || [];
		const selectedLemmaEdition =
			this.data.lemmaText[this.state.selectedLemmaEditionIndex] || { lines: [] };
		const styles = {
			lemmaReferenceModal: {
				top: this.props.top,
				left: this.props.left,

			},

		};
		const hasLemma = ~selectedLemmaEdition.lines.length;

		return (
			<div
				className={`reference-modal${(this.props.visible && hasLemma) ?
					' reference-modal-visible' : ''}`}
				style={styles.lemmaReferenceModal}
			>
				<article className="comment	lemma-comment paper-shadow ">

					<div className="reference-text">
						{selectedLemmaEdition.lines.map((line, i) => (
							<p
								key={i}
								className="lemma-text"
								dangerouslySetInnerHTML={{ __html: line.html }}
							/>
						))}
					</div>

					<div className="edition-tabs tabs">
						{lemmaText.map((lemmaTextEdition, i) => {
							const lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);
							return (
								<RaisedButton
									key={i}
									label={lemmaEditionTitle}
									data-edition={lemmaTextEdition.title}
									className={selectedLemmaEdition.slug === lemmaTextEdition.slug ?
										'edition-tab tab selected-edition-tab' : 'edition-tab tab'}
									onClick={self.toggleEdition.bind(null, lemmaTextEdition.slug)}
								/>
							);
						})}
					</div>
					<div className="meta-tabs tabs">
						{/*
						<FlatButton
							label="Entities"
							className="edition-tab tab"
							onClick={this.toggleEntities}
						/>
						<FlatButton
							label="Scansion"
							className="edition-tab tab"
							onClick={this.toggleScansion}
						/>
						*/}
					</div>

					<i
						className="mdi mdi-close paper-shadow"
						onClick={this.props.closeLemmaReference}
					/>
				</article>

			</div>

		);
	},

});

export default LemmaReferenceModal;
