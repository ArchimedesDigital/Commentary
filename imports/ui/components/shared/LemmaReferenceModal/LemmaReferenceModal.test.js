import React from 'react';
import renderer from 'react-test-renderer';

// component:
import LemmaReferenceModal from './LemmaReferenceModal';

describe('LemmaReferenceModal', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(
				<LemmaReferenceModal
					top={1}
					left={1}
					work="testWork"
					subwork={1}
					lineFrom={1}
					closeLemmaReference={() => {}}
					lemmaText={[]}
				/>)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

// TODO LemmaReference needs to be refactored: get rid of mix-ins and use container


