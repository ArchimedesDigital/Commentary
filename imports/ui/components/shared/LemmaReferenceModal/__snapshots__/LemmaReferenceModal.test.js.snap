// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LemmaReferenceModal renders correctly 1`] = `
<div
  className="reference-modal react-draggable"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "left": 1,
      "top": 1,
      "touchAction": "none",
      "transform": "translate(0px,0px)",
    }
  }
  transform={null}
>
  <article
    className="comment	lemma-comment paper-shadow "
  >
    <div
      className="reference-text"
    >
      <div
        className="loading-mock lemma-filler lemma-filler-1"
      />
      <div
        className="loading-mock lemma-filler lemma-filler-2"
      />
    </div>
    <div
      className="edition-tabs tabs"
    />
    <div
      className="meta-tabs tabs"
    />
    <i
      className="mdi mdi-close paper-shadow"
      onClick={[Function]}
    />
  </article>
</div>
`;
